container_linux.go

```diff
@@ -252,6 +255,338 @@ func (c *linuxContainer) NotifyOOM() (<-chan struct{}, error) {
 }
 
+// XXX debug support, remove when debugging done.
+func addArgsFromEnv(evar string, args *[]string) {
+	if e := os.Getenv(evar); e != "" {
+		for _, f := range strings.Fields(e) {
+			*args = append(*args, f)
+		}
+	}
+	fmt.Printf(">>> criu %v\n", *args)
+}
+
+func (c *linuxContainer) checkCriuVersion() error {
+	var x, y, z int
+
+	out, err := exec.Command(c.criuPath, "-V").Output()
+	if err != nil {
+		return err
+	}
+
+	n, err := fmt.Sscanf(string(out), "Version: %d.%d.%d", &x, &y, &z)
+	if n < 2 || err != nil {
+		return fmt.Errorf("Unable to parse the CRIU version: %s", out)
+	}
+
+	if x*10000+y*100+z < 10501 {
+		return fmt.Errorf("CRIU version must be 1.5.1 or higher")
+	}
+
+	return nil
+}
+
+func (c *linuxContainer) Checkpoint(imagePath string) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if err := c.checkCriuVersion(); err != nil {
+		return err
+	}
+
+	workPath := filepath.Join(c.root, "criu.work")
+	if err := os.Mkdir(workPath, 0655); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	workDir, err := os.Open(workPath)
+	if err != nil {
+		return err
+	}
+	defer workDir.Close()
+
+	imageDir, err := os.Open(imagePath)
+	if err != nil {
+		return err
+	}
+	defer imageDir.Close()
+	t := criurpc.CriuReqType_DUMP
+	req := criurpc.CriuReq{
+		Type: &t,
+		Opts: &criurpc.CriuOpts{
+			ImagesDirFd:   proto.Int32(int32(imageDir.Fd())),
+			WorkDirFd:     proto.Int32(int32(workDir.Fd())),
+			LogLevel:      proto.Int32(4),
+			LogFile:       proto.String("dump.log"),
+			Root:          proto.String(c.config.Rootfs),
+			ManageCgroups: proto.Bool(true),
+			NotifyScripts: proto.Bool(true),
+			Pid:           proto.Int32(int32(c.initProcess.pid())),
+		},
+	}
+	for _, m := range c.config.Mounts {
+		if m.Device == "bind" {
+			mountDest := m.Destination
+			if strings.HasPrefix(mountDest, c.config.Rootfs) {
+				mountDest = mountDest[len(c.config.Rootfs):]
+			}
+
+			extMnt := new(criurpc.ExtMountMap)
+			extMnt.Key = proto.String(mountDest)
+			extMnt.Key = proto.String(m.Destination)
+			extMnt.Val = proto.String(m.Destination)
+			req.Opts.ExtMnt = append(req.Opts.ExtMnt, extMnt)
+		}
+	}
+
+	err = c.criuSwrk(nil, &req, imagePath)
+	if err != nil {
+		log.Errorf(filepath.Join(workPath, "dump.log"))
+		return err
+	}
+
+	log.Info("Checkpointed")
+	return nil
+}
+
+func (c *linuxContainer) Restore(process *Process, imagePath string) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if err := c.checkCriuVersion(); err != nil {
+		return err
+	}
+
+	workPath := filepath.Join(c.root, "criu.work")
+	// Since a container can be C/R'ed multiple times,
+	// the work directory may already exist.
+	if err := os.Mkdir(workPath, 0755); err != nil && !os.IsExist(err) {
+		return err
+	}
+	workDir, err := os.Open(workPath)
+	if err != nil {
+		return err
+	}
+	defer workDir.Close()
+
+	imageDir, err := os.Open(imagePath)
+	if err != nil {
+		return err
+	}
+	defer imageDir.Close()
+
+	root := filepath.Join(c.root, "criu-root")
```

Comment:
@boucher CRIU has a few requirements for a root directory:
* it must be a mount point
* its parent must not be overmounted

We create a temporary directory only to satisfy these requirements.